# Analyze A/B Test Results 

This project will assure you have mastered the subjects covered in the statistics lessons. We have organized the current notebook into the following sections: 

- [Introduction](#intro)
- [Part I - Probability](#probability)
- [Part II - A/B Test](#ab_test)
- [Part III - Regression](#regression)
- [Final Check](#finalcheck)
- [Submission](#submission)

Specific programming tasks are marked with a **ToDo** tag. 

<a id='intro'></a>
## Introduction

A/B tests are very commonly performed by data analysts and data scientists. For this project, you will be working to understand the results of an A/B test run by an e-commerce website.  Your goal is to work through this notebook to help the company understand if they should:
- Implement the new webpage, 
- Keep the old webpage, or 
- Perhaps run the experiment longer to make their decision.

Each **ToDo** task below has an associated quiz present in the classroom.  Though the classroom quizzes are **not necessary** to complete the project, they help ensure you are on the right track as you work through the project, and you can feel more confident in your final submission meeting the [rubric](https://review.udacity.com/#!/rubrics/1214/view) specification. 

>**Tip**: Though it's not a mandate, students can attempt the classroom quizzes to ensure statistical numeric values are calculated correctly in many cases.

<a id='probability'></a>
## Part I - Probability

To get started, let's import our libraries.

In [1]:
import pandas as pd
import numpy as np
import random
import matplotlib.pyplot as plt
%matplotlib inline
import statsmodels.api as sm
#We are setting the seed to assure you get the same answers on quizzes as we set up
random.seed(42)

/opt/conda/lib/python3.6/site-packages/statsmodels/compat/pandas.py:56: FutureWarning: The pandas.core.datetools module is deprecated and will be removed in a future version. Please use the pandas.tseries module instead.
  from pandas.core import datetools


### ToDo 1.1
Now, read in the `ab_data.csv` data. Store it in `df`. Below is the description of the data, there are a total of 5 columns:

<center>

|Data columns|Purpose|Valid values|
| ------------- |:-------------| -----:|
|user_id|Unique ID|Int64 values|
|timestamp|Time stamp when the user visited the webpage|-|
|group|In the current A/B experiment, the users are categorized into two broad groups. <br>The `control` group users are expected to be served with `old_page`; and `treatment` group users are matched with the `new_page`. <br>However, **some inaccurate rows** are present in the initial data, such as a `control` group user is matched with a `new_page`. |`['control', 'treatment']`|
|landing_page|It denotes whether the user visited the old or new webpage.|`['old_page', 'new_page']`|
|converted|It denotes whether the user decided to pay for the company's product. Here, `1` means yes, the user bought the product.|`[0, 1]`|
</center>
Use your dataframe to answer the questions in Quiz 1 of the classroom.


>**Tip**: Please save your work regularly.

**a.** Read in the dataset from the `ab_data.csv` file and take a look at the top few rows here:

In [2]:
df = pd.read_csv('ab_data.csv')
df.head()

user_id                   timestamp      group landing_page  converted
0   851104  2017-01-21 22:11:48.556739    control     old_page          0
1   804228  2017-01-12 08:01:45.159739    control     old_page          0
2   661590  2017-01-11 16:55:06.154213  treatment     new_page          0
3   853541  2017-01-08 18:28:03.143765  treatment     new_page          0
4   864975  2017-01-21 01:52:26.210827    control     old_page          1

**b.** Use the cell below to find the number of rows in the dataset.

In [3]:
df.shape[0]

294478

**c.** The number of unique users in the dataset.

In [4]:
df['user_id'].nunique()

290584

**d.** The proportion of users converted.

In [5]:
(df['converted'] == 1).mean()

0.11965919355605512

In [6]:
(df.query('group == "treatment" and landing_page != "new_page" or group == "control" and landing_page == "new_page"')).shape[0]


3893

In [7]:
treatment_df = df.query('group == "treatment"')


**f.** Do any of the rows have missing values?

In [8]:
df.isnull().sum()

user_id         0
timestamp       0
group           0
landing_page    0
converted       0
dtype: int64

### ToDo 1.2  
In a particular row, the **group** and **landing_page** columns should have either of the following acceptable values:

|user_id| timestamp|group|landing_page|converted|
|---|---|---|---|---|
|XXXX|XXXX|`control`| `old_page`|X |
|XXXX|XXXX|`treatment`|`new_page`|X |


It means, the `control` group users should match with `old_page`; and `treatment` group users should matched with the `new_page`. 

However, for the rows where `treatment` does not match with `new_page` or `control` does not match with `old_page`, we cannot be sure if such rows truly received the new or old wepage.  


Use **Quiz 2** in the classroom to figure out how should we handle the rows where the group and landing_page columns don't match?

**a.** Now use the answer to the quiz to create a new dataset that meets the specifications from the quiz.  Store your new dataframe in **df2**.

In [9]:
df_unacurate = df.query("(group == 'treatment' and landing_page == 'old_page')")

In [10]:
# Remove the inaccurate rows, and store the result in a new dataframe df2
df2 = df.drop(df_unacurate.index)

In [11]:
# Double Check all of the incorrect rows were removed from df2 - 
# Output of the statement below should be 0
df2[((df2['group'] == 'treatment') == (df2['landing_page'] == 'new_page')) == False].shape[0]

1928

### ToDo 1.3  
Use **df2** and the cells below to answer questions for **Quiz 3** in the classroom.

**a.** How many unique **user_id**s are in **df2**?

In [12]:
df2.user_id.nunique()

290584

**b.** There is one **user_id** repeated in **df2**.  What is it?

In [13]:
df.isnull().sum()

user_id         0
timestamp       0
group           0
landing_page    0
converted       0
dtype: int64

**c.** Display the rows for the duplicate **user_id**? 

In [15]:
dublicated = df2[df2.user_id.duplicated(keep=False) ]
dublicated

user_id                   timestamp      group landing_page  converted
22       767017  2017-01-12 22:58:14.991443    control     new_page          0
226      773693  2017-01-23 18:05:45.167335    control     old_page          1
240      733976  2017-01-11 15:11:16.407599    control     new_page          0
303      753749  2017-01-06 01:30:10.493842    control     old_page          0
490      808613  2017-01-10 21:44:01.292755    control     new_page          0
654      823319  2017-01-20 09:02:25.852683  treatment     new_page          0
655      726766  2017-01-16 11:00:53.912376  treatment     new_page          0
753      646066  2017-01-04 19:04:49.423564  treatment     new_page          1
774      864223  2017-01-03 18:45:03.638277  treatment     new_page          0
846      637639  2017-01-11 23:09:52.682329    control     new_page          1
850      793580  2017-01-08 03:25:33.723712    control     new_page          1
988      698120  2017-01-22 07:09:37.540970    control     new_page          0
1198     646342  2017-01-06 18:39:23.484797    control     new_page          0
1252     883950  2017-01-07 11:53:42.459470    control     old_page          0
1282     630780  2017-01-09 10:14:27.657854    control     old_page          0
1335     777170  2017-01-11 20:19:36.335726  treatment     new_page          1
1354     735021  2017-01-16 09:51:29.349493    control     new_page          0
1361     785823  2017-01-04 09:48:53.753839  treatment     new_page          0
1474     678638  2017-01-18 06:36:42.515395    control     new_page          0
1662     766516  2017-01-19 05:45:06.467040    control     old_page          0
1716     910059  2017-01-12 14:01:45.357784  treatment     new_page          0
1836     792397  2017-01-18 10:22:06.674044  treatment     new_page          0
1877     717682  2017-01-07 03:05:39.891873    control     new_page          0
1899     773192  2017-01-09 05:37:58.781806  treatment     new_page          0
2023     937692  2017-01-19 01:29:42.739007    control     new_page          0
2056     642390  2017-01-23 14:29:44.728785  treatment     new_page          0
2074     852848  2017-01-20 11:56:05.325544    control     old_page          0
2105     817933  2017-01-14 22:19:35.027511  treatment     new_page          1
2118     738864  2017-01-14 15:51:31.645383  treatment     new_page          0
2214     649781  2017-01-20 03:50:20.837704    control     new_page          0
...         ...                         ...        ...          ...        ...
292388   920529  2017-01-19 21:03:51.381237  treatment     new_page          0
292405   819974  2017-01-03 05:58:44.734645    control     new_page          0
292433   837359  2017-01-07 05:10:48.213079  treatment     new_page          0
292499   787405  2017-01-04 08:22:12.854996    control     old_page          0
292562   709281  2017-01-03 01:10:41.185849    control     old_page          0
292570   778969  2017-01-21 12:59:42.740399    control     new_page          1
292585   682832  2017-01-03 07:09:30.004752    control     old_page          0
292718   817824  2017-01-24 09:14:52.929636  treatment     new_page          0
292748   684361  2017-01-19 03:59:57.656614    control     new_page          0
292845   893018  2017-01-10 15:05:37.522921    control     new_page          0
293017   792268  2017-01-06 09:21:58.341063    control     new_page          0
293085   884635  2017-01-19 14:19:48.484389    control     new_page          0
293099   869470  2017-01-03 10:05:40.466386    control     old_page          0
293298   762128  2017-01-11 23:10:47.420797    control     old_page          0
293393   636565  2017-01-12 07:26:31.103374    control     new_page          0
293449   715367  2017-01-13 09:06:08.632332  treatment     new_page          0
293480   638376  2017-01-18 15:41:02.395882    control     new_page          0
293568   704024  2017-01-15 17:06:09.309987    control     new_page          0
293604   648354  2017-01-20 01:51:29.007764    control    

**d.** Remove **one** of the rows with a duplicate **user_id**, from the **df2** dataframe.

In [16]:
# Remove one of the rows with a duplicate user_id..
# Hint: The dataframe.drop_duplicates() may not work in this case because the rows with duplicate user_id are not entirely identical. 
df2 = df2.drop(2893)
# Check again if the row with a duplicate user_id is deleted or not
df2[df2.user_id == 773192]

user_id                   timestamp      group landing_page  converted
1899   773192  2017-01-09 05:37:58.781806  treatment     new_page          0

In [17]:
df2.user_id.duplicated().sum()

1928

### ToDo 1.4  
Use **df2** in the cells below to answer the quiz questions related to **Quiz 4** in the classroom.

**a.** What is the probability of an individual converting regardless of the page they receive?<br><br>

>**Tip**: The probability  you'll compute represents the overall "converted" success rate in the population and you may call it $p_{population}$.



In [18]:
df2.converted.mean()

0.11960876818728804

**b.** Given that an individual was in the `control` group, what is the probability they converted?

In [19]:
prob_control = (df2.query('group == "control"')['converted']).mean()
prob_control

0.12039917935897611

**c.** Given that an individual was in the `treatment` group, what is the probability they converted?

In [20]:
prob_treatment = (df2.query('group == "treatment"')['converted']).mean()
prob_treatment

0.11880806551510564

>**Tip**: The probabilities you've computed in the points (b). and (c). above can also be treated as conversion rate. 
Calculate the actual difference  (`obs_diff`) between the conversion rates for the two groups. You will need that later.  

In [21]:
# Calculate the actual difference (obs_diff) between the conversion rates for the two groups.
obs_diff = prob_treatment - prob_control
obs_diff

-0.0015911138438704647

**d.** What is the probability that an individual received the new page?

In [22]:
(df2['landing_page'] == 'new_page').mean()

0.50335712722896842

**e.** Consider your results from parts (a) through (d) above, and explain below whether the new `treatment` group users lead to more conversions.

>**The performance is similar on both pages, so there is no evidence of conversion.**

<a id='ab_test'></a>
## Part II - A/B Test

Since a timestamp is associated with each event, you could run a hypothesis test continuously as long as you observe the events. 

However, then the hard questions would be: 
- Do you stop as soon as one page is considered significantly better than another or does it need to happen consistently for a certain amount of time?  
- How long do you run to render a decision that neither page is better than another?  

These questions are the difficult parts associated with A/B tests in general.  


### ToDo 2.1
For now, consider you need to make the decision just based on all the data provided.  

> Recall that you just calculated that the "converted" probability (or rate) for the old page is *slightly* higher than that of the new page (ToDo 1.4.c). 

If you want to assume that the old page is better unless the new page proves to be definitely better at a Type I error rate of 5%, what should be your null and alternative hypotheses (**$H_0$** and **$H_1$**)?  

You can state your hypothesis in terms of words or in terms of **$p_{old}$** and **$p_{new}$**, which are the "converted" probability (or rate) for the old and new pages respectively.

- **$H_0$** : p_old ≥ p_new
- **$H_1$** : p_old < p_new

### ToDo 2.2 - Null Hypothesis $H_0$ Testing
Under the null hypothesis $H_0$, assume that $p_{new}$ and $p_{old}$ are equal. Furthermore, assume that $p_{new}$ and $p_{old}$ both are equal to the **converted** success rate in the `df2` data regardless of the page. So, our assumption is: <br><br>
<center>
$p_{new}$ = $p_{old}$ = $p_{population}$
</center>

In this section, you will: 

- Simulate (bootstrap) sample data set for both groups, and compute the  "converted" probability $p$ for those samples. 


- Use a sample size for each group equal to the ones in the `df2` data.


- Compute the difference in the "converted" probability for the two samples above. 


- Perform the sampling distribution for the "difference in the converted probability" between the two simulated-samples over 10,000 iterations; and calculate an estimate. 



Use the cells below to provide the necessary parts of this simulation.  You can use **Quiz 5** in the classroom to make sure you are on the right track.

**a.** What is the **conversion rate** for $p_{new}$ under the null hypothesis? 

In [23]:
p_new = df2.converted.mean()
p_new

0.11960876818728804

**b.** What is the **conversion rate** for $p_{old}$ under the null hypothesis? 

In [24]:
p_old = df2.converted.mean()
p_old

0.11960876818728804

**c.** What is $n_{new}$, the number of individuals in the treatment group? <br><br>
*Hint*: The treatment group users are shown the new page.

In [25]:
n_new = (df2.query('landing_page == "new_page"')['group'] == 'treatment').sum()
n_new

145310

**d.** What is $n_{old}$, the number of individuals in the control group?

In [26]:
n_old = (df2.query('landing_page == "old_page"')['group'] == 'control').sum()
n_old

145274

**e. Simulate Sample for the `treatment` Group**<br> 
Simulate $n_{new}$ transactions with a conversion rate of $p_{new}$ under the null hypothesis.  <br><br>
*Hint*: Use `numpy.random.choice()` method to randomly generate $n_{new}$ number of values. <br>
Store these $n_{new}$ 1's and 0's in the `new_page_converted` numpy array.


In [27]:
# Simulate a Sample for the treatment Group
new_page_converted = (np.random.choice([0,1],size = n_new , p =[(1 - p_new),p_new ])).mean()
new_page_converted

0.12114100887757209

**f. Simulate Sample for the `control` Group** <br>
Simulate $n_{old}$ transactions with a conversion rate of $p_{old}$ under the null hypothesis. <br> Store these $n_{old}$ 1's and 0's in the `old_page_converted` numpy array.

In [28]:
# Simulate a Sample for the control Group
old_page_converted = (np.random.choice([0,1],size = n_old , p =[(1 - p_old), p_old])).mean()
old_page_converted

0.12064099563583298

**g.** Find the difference in the "converted" probability $(p{'}_{new}$ - $p{'}_{old})$ for your simulated samples from the parts (e) and (f) above. 

In [29]:
diff = new_page_converted - old_page_converted
diff

0.00050001324173910844


**h. Sampling distribution** <br>
Re-create `new_page_converted` and `old_page_converted` and find the $(p{'}_{new}$ - $p{'}_{old})$ value 10,000 times using the same simulation process you used in parts (a) through (g) above. 

<br>
Store all  $(p{'}_{new}$ - $p{'}_{old})$  values in a NumPy array called `p_diffs`.

In [30]:
# Sampling distribution 
p_diffs = []
for i in range(int(1e4)):
    new_page_converted = (np.random.choice([0,1],size = n_new , p =[(1 - p_new),p_new ])).mean()
    old_page_converted = (np.random.choice([0,1],size = n_old , p =[(1 - p_old), p_old])).mean()
    p_diffs.append(new_page_converted - old_page_converted)


In [31]:
# convert p_diffs to array
p_diffs = np.array(p_diffs)

**i. Histogram**<br> 
Plot a histogram of the **p_diffs**.  Does this plot look like what you expected?  Use the matching problem in the classroom to assure you fully understand what was computed here.<br><br>

Also, use `plt.axvline()` method to mark the actual difference observed  in the `df2` data (recall `obs_diff`), in the chart.  

>**Tip**: Display title, x-label, and y-label in the chart.

In [32]:
!pip install plotly
#!pip install chart_studio
#!pip install cufflinks
!pip install plotly_express
!pip install --upgrade plotly 

%load_ext autoreload
%autoreload 2

    100% |████████████████████████████████| 235kB 7.9MB/s eta 0:00:01
    100% |████████████████████████████████| 15.2MB 2.4MB/s eta 0:00:01  4% |█▌                              | 686kB 27.0MB/s eta 0:00:01    34% |███████████▏                    | 5.3MB 31.5MB/s eta 0:00:01    73% |███████████████████████▋        | 11.2MB 33.5MB/s eta 0:00:01
    100% |████████████████████████████████| 9.5MB 3.6MB/s eta 0:00:01    28% |█████████▎                      | 2.7MB 32.6MB/s eta 0:00:01    58% |██████████████████▋             | 5.5MB 30.0MB/s eta 0:00:01
statsmodels 0.12.2 has requirement numpy>=1.15, but you'll have numpy 1.12.1 which is incompatible.
  Found existing installation: patsy 0.4.1
    Uninstalling patsy-0.4.1:
      Successfully uninstalled patsy-0.4.1
  Found existing installation: plotly 2.0.15
    Uninstalling plotly-2.0.15:
      Successfully uninstalled plotly-2.0.15
  Found existing installation: statsmodels 0.8.0
    Uninstalling statsmodels-0.8.0:
      Successfully unin

In [37]:
#import plotly.figure_factory as ff
#import chart_studio.plotly as py
import plotly.express as px


In [38]:
fig = px.histogram(p_diffs)
fig.add_vline(x=obs_diff, line_width=3, line_dash="dash", line_color="red")
fig.show()

**j.** What proportion of the **p_diffs** are greater than the actual difference observed in the `df2` data?

In [39]:
(p_diffs > obs_diff).mean()

0.90820000000000001

**k.** Please explain in words what you have just computed in part **j** above.  
 - What is this value called in scientific studies?  
 - What does this value signify in terms of whether or not there is a difference between the new and old pages? *Hint*: Compare the value above with the "Type I error rate (0.05)". 

- >**P-value**
- >**we can't reject null hypothesis becouse P-value more than 0.5**




**l. Using Built-in Methods for Hypothesis Testing**<br>
We could also use a built-in to achieve similar results.  Though using the built-in might be easier to code, the above portions are a walkthrough of the ideas that are critical to correctly thinking about statistical significance. 

Fill in the statements below to calculate the:
- `convert_old`: number of conversions with the old_page
- `convert_new`: number of conversions with the new_page
- `n_old`: number of individuals who were shown the old_page
- `n_new`: number of individuals who were shown the new_page


In [40]:
import statsmodels.api as sm

# number of conversions with the old_page
convert_old = (df2.query('landing_page== "old_page"')['converted']).sum()
print(convert_old)
# number of conversions with the new_page
convert_new = (df2.query('landing_page== "new_page"')['converted']).sum()
print(convert_new)
# number of individuals who were shown the old_page
n_old = (df2['landing_page'] == "old_page").sum()
print(n_old)
# number of individuals who received new_page
n_new = (df2['landing_page'] == "new_page").sum()
print(n_new)

17489
17498
145274
147238


**m.** Now use `sm.stats.proportions_ztest()` to compute your test statistic and p-value.  [Here](https://www.statsmodels.org/stable/generated/statsmodels.stats.proportion.proportions_ztest.html) is a helpful link on using the built in.

The syntax is: 
```bash
proportions_ztest(count_array, nobs_array, alternative='larger')
```
where, 
- `count_array` = represents the number of "converted" for each group
- `nobs_array` = represents the total number of observations (rows) in each group
- `alternative` = choose one of the values from `[‘two-sided’, ‘smaller’, ‘larger’]` depending upon two-tailed, left-tailed, or right-tailed respectively. 
>**Hint**: <br>
It's a two-tailed if you defined $H_1$ as $(p_{new} = p_{old})$. <br>
It's a left-tailed if you defined $H_1$ as $(p_{new} < p_{old})$. <br>
It's a right-tailed if you defined $H_1$ as $(p_{new} > p_{old})$. 

The built-in function above will return the z_score, p_value. 

---
### About the two-sample z-test
Recall that you have plotted a distribution `p_diffs` representing the
difference in the "converted" probability  $(p{'}_{new}-p{'}_{old})$  for your two simulated samples 10,000 times. 

Another way for comparing the mean of two independent and normal distribution is a **two-sample z-test**. You can perform the Z-test to calculate the Z_score, as shown in the equation below:

$$
Z_{score} = \frac{ (p{'}_{new}-p{'}_{old}) - (p_{new}  -  p_{old})}{ \sqrt{ \frac{\sigma^{2}_{new} }{n_{new}} + \frac{\sigma^{2}_{old} }{n_{old}}  } }
$$

where,
- $p{'}$ is the "converted" success rate in the sample
- $p_{new}$ and $p_{old}$ are the "converted" success rate for the two groups in the population. 
- $\sigma_{new}$ and $\sigma_{new}$ are the standard deviation for the two groups in the population. 
- $n_{new}$ and $n_{old}$ represent the size of the two groups or samples (it's same in our case)


>Z-test is performed when the sample size is large, and the population variance is known. The z-score represents the distance between the two "converted" success rates in terms of the standard error. 

Next step is to make a decision to reject or fail to reject the null hypothesis based on comparing these two values: 
- $Z_{score}$
- $Z_{\alpha}$ or $Z_{0.05}$, also known as critical value at 95% confidence interval.  $Z_{0.05}$ is 1.645 for one-tailed tests,  and 1.960 for two-tailed test. You can determine the $Z_{\alpha}$ from the z-table manually. 

Decide if your hypothesis is either a two-tailed, left-tailed, or right-tailed test. Accordingly, reject OR fail to reject the  null based on the comparison between $Z_{score}$ and $Z_{\alpha}$. 
>Hint:<br>
For a right-tailed test, reject null if $Z_{score}$ > $Z_{\alpha}$. <br>
For a left-tailed test, reject null if $Z_{score}$ < $Z_{\alpha}$. 


In other words, we determine whether or not the $Z_{score}$ lies in the "rejection region" in the distribution. A "rejection region" is an interval where the null hypothesis is rejected iff the $Z_{score}$ lies in that region.



Reference: 
- Example 9.1.2 on this [page](https://stats.libretexts.org/Bookshelves/Introductory_Statistics/Book%3A_Introductory_Statistics_(Shafer_and_Zhang)/09%3A_Two-Sample_Problems/9.01%3A_Comparison_of_Two_Population_Means-_Large_Independent_Samples), courtesy www.stats.libretexts.org

---

>**Tip**: You don't have to dive deeper into z-test for this exercise. **Try having an overview of what does z-score signify in general.** 

In [42]:
import statsmodels.api as sm
# ToDo: Complete the sm.stats.proportions_ztest() method arguments
count = np.array([17489, 17264])
nobs = np.array([145274, 145310])
z_score, p_value = sm.stats.proportions_ztest(count, nobs, alternative = 'smaller')
print(z_score, p_value)

1.31092419842 0.905058312759


**n.** What do the z-score and p-value you computed in the previous question mean for the conversion rates of the old and new pages?  Do they agree with the findings in parts **j.** and **k.**?<br><br>

>**Tip**: Notice whether the p-value is similar to the one computed earlier. Accordingly, can you reject/fail to reject the null hypothesis? It is important to correctly interpret the test statistic and p-value.

- >**z_score is 1.31 and confidence interval is 1.960 Hence, we can't reject Null Hypothesis.**
- >**p-value agree with the finding in part j and k.**

<a id='regression'></a>
### Part III - A regression approach

### ToDo 3.1 
In this final part, you will see that the result you achieved in the A/B test in Part II above can also be achieved by performing regression.<br><br> 

**a.** Since each row in the `df2` data is either a conversion or no conversion, what type of regression should you be performing in this case?

>**Logistic regression.**

**b.** The goal is to use **statsmodels** library to fit the regression model you specified in part **a.** above to see if there is a significant difference in conversion based on the page-type a customer receives. However, you first need to create the following two columns in the `df2` dataframe:
 1. `intercept` - It should be `1` in the entire column. 
 2. `ab_page` - It's a dummy variable column, having a value `1` when an individual receives the **treatment**, otherwise `0`.  

In [43]:
df2.head()

user_id                   timestamp      group landing_page  converted
0   851104  2017-01-21 22:11:48.556739    control     old_page          0
1   804228  2017-01-12 08:01:45.159739    control     old_page          0
2   661590  2017-01-11 16:55:06.154213  treatment     new_page          0
3   853541  2017-01-08 18:28:03.143765  treatment     new_page          0
4   864975  2017-01-21 01:52:26.210827    control     old_page          1

In [44]:
import statsmodels.api as sm

In [45]:
df2['intercept'] = 1
df2['ab_page'] = pd.get_dummies(df['group'])['control']

**c.** Use **statsmodels** to instantiate your regression model on the two columns you created in part (b). above, then fit the model to predict whether or not an individual converts. 


In [46]:
logistic = sm.Logit(df2['converted'], df2[['intercept', 'ab_page']])
result = logistic.fit()

Optimization terminated successfully.
         Current function value: 0.366142
         Iterations 6


**d.** Provide the summary of your model below, and use it as necessary to answer the following questions.

In [47]:
result.summary2()

<class 'statsmodels.iolib.summary2.Summary'>
"""
                          Results: Logit
==================================================================
Model:              Logit            No. Iterations:   6.0000     
Dependent Variable: converted        Pseudo R-squared: 0.000      
Date:               2022-07-26 11:39 AIC:              214205.7162
No. Observations:   292512           BIC:              214226.8887
Df Model:           1                Log-Likelihood:   -1.0710e+05
Df Residuals:       292510           LL-Null:          -1.0710e+05
Converged:          1.0000           Scale:            1.0000     
-------------------------------------------------------------------
              Coef.   Std.Err.      z      P>|z|    [0.025   0.975]
-------------------------------------------------------------------
intercept    -2.0038    0.0081  -247.1457  0.0000  -2.0197  -1.9879
ab_page       0.0151    0.0114     1.3259  0.1849  -0.0072   0.0374
==================================================================

"""

**e.** What is the p-value associated with **ab_page**? Why does it differ from the value you found in **Part II**?<br><br>  

**Hints**: 
- What are the null and alternative hypotheses associated with your regression model, and how do they compare to the null and alternative hypotheses in **Part II**? 
- You may comment on if these hypothesis (Part II vs. Part III) are one-sided or two-sided. 
- You may also compare the current p-value with the Type I error rate (0.05).


- **$H_0$ : page type has no effect on conversion rate**  
- p_new - p_old ≤ 0
- **$H_1$ : Page type causes a variationin conversion rate**  
- p_new - p_old > 0

**f.** Now, you are considering other things that might influence whether or not an individual converts.  Discuss why it is a good idea to consider other factors to add into your regression model.  Are there any disadvantages to adding additional terms into your regression model?

>**Factors that can be added to the model.**
- User age
- user gender
- page colors


**g. Adding countries**<br> 
Now along with testing if the conversion rate changes for different pages, also add an effect based on which country a user lives in. 

1. You will need to read in the **countries.csv** dataset and merge together your `df2` datasets on the appropriate rows. You call the resulting dataframe `df_merged`. [Here](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.join.html) are the docs for joining tables. 

2. Does it appear that country had an impact on conversion?  To answer this question, consider the three unique values, `['UK', 'US', 'CA']`, in the `country` column. Create dummy variables for these country columns. 
>**Hint:** Use `pandas.get_dummies()` to create dummy variables. **You will utilize two columns for the three dummy variables.** 

 Provide the statistical output as well as a written response to answer this question.

In [48]:
# Read the countries.csv
df_countries = pd.read_csv('countries.csv')
df_countries.head()

user_id country
0   834778      UK
1   928468      US
2   822059      UK
3   711597      UK
4   710616      UK

In [49]:
# Join with the df2 dataframe
df2 = pd.merge(df2, df_countries, on='user_id')

In [50]:
# Create the necessary dummy variables
df2[['ca','uk','us']] = pd.get_dummies(df2['country'])

**h. Fit your model and obtain the results**<br> 
Though you have now looked at the individual factors of country and page on conversion, we would now like to look at an interaction between page and country to see if are there significant effects on conversion.  **Create the necessary additional columns, and fit the new model.** 


Provide the summary results (statistical output), and your conclusions (written response) based on the results. 

>**Tip**: Conclusions should include both statistical reasoning, and practical reasoning for the situation. 

>**Hints**: 
- Look at all of p-values in the summary, and compare against the Type I error rate (0.05). 
- Can you reject/fail to reject the null hypotheses (regression model)?
- Comment on the effect of page and country to predict the conversion.


In [51]:
df2['UK_ab_page'] = df2['uk'] * df2['ab_page']
df2['CA_ab_page'] = df2['ca'] * df2['ab_page']

In [54]:
# Fit your model, and summarize the results
logistic2 = sm.Logit(df2['converted'], df2[['intercept', 'ab_page','ca','uk', 'UK_ab_page', 'CA_ab_page']])
result2 = logistic2.fit()
result2.summary2()

Optimization terminated successfully.
         Current function value: 0.366132
         Iterations 6


<class 'statsmodels.iolib.summary2.Summary'>
"""
                          Results: Logit
==================================================================
Model:              Logit            No. Iterations:   6.0000     
Dependent Variable: converted        Pseudo R-squared: 0.000      
Date:               2022-07-26 11:40 AIC:              214208.1803
No. Observations:   292512           BIC:              214271.6979
Df Model:           5                Log-Likelihood:   -1.0710e+05
Df Residuals:       292506           LL-Null:          -1.0710e+05
Converged:          1.0000           Scale:            1.0000     
-------------------------------------------------------------------
              Coef.   Std.Err.      z      P>|z|    [0.025   0.975]
-------------------------------------------------------------------
intercept    -2.0070    0.0097  -207.0454  0.0000  -2.0260  -1.9880
ab_page       0.0198    0.0136     1.4561  0.1454  -0.0069   0.0465
ca           -0.0644    0.0384    -1.6788  0.0932  -0.1396   0.0108
uk            0.0257    0.0188     1.3634  0.1728  -0.0112   0.0625
UK_ab_page   -0.0291    0.0265    -1.0964  0.2729  -0.0810   0.0229
CA_ab_page    0.0524    0.0535     0.9782  0.3280  -0.0526   0.1573
==================================================================

"""

>## conclusion.
- **The old page is better than the new page**
- **We reject the Alternate Hypothesis**
- **We accept the Null Hypothesis**
- **There is no clear evidence that the state is an influencer**


<a id='finalcheck'></a>
## Final Check!

Congratulations!  You have reached the end of the A/B Test Results project!  You should be very proud of all you have accomplished!

> **Tip**: Once you are satisfied with your work here, check over your notebook to make sure that it satisfies all the specifications mentioned in the rubric. You should also probably remove all of the "Hints" and "Tips" like this one so that the presentation is as polished as possible.


<a id='submission'></a>
## Submission
You may either submit your notebook through the "SUBMIT PROJECT" button at the bottom of this workspace, or you may work from your local machine and submit on  the last page of this project lesson.  

1. Before you submit your project, you need to create a .html or .pdf version of this notebook in the workspace here. To do that, run the code cell below. If it worked correctly, you should get a return code of 0, and you should see the generated .html file in the workspace directory (click on the orange Jupyter icon in the upper left).


2. Alternatively, you can download this report as .html via the **File** > **Download as** submenu, and then manually upload it into the workspace directory by clicking on the orange Jupyter icon in the upper left, then using the Upload button.


3. Once you've done this, you can submit your project by clicking on the "Submit Project" button in the lower right here. This will create and submit a zip file with this .ipynb doc and the .html or .pdf version you created. Congratulations!

In [55]:
from subprocess import call
call(['python', '-m', 'nbconvert', 'Analyze_ab_test_results_notebook.ipynb'])

0